fri/src/folder.rs

```diff
@@ -1,10 +1,41 @@
+use anyhow::Error;
 use ark_ff::{FftField, PrimeField};
+use ark_poly::Radix2EvaluationDomain;
 use std::sync::Arc;
 
+use crate::stone_domain::get_field_element_at_index;
+
 pub struct MultiplicativeFriFolder;
 
 #[allow(dead_code)]
 impl MultiplicativeFriFolder {
+    // Computes the values of the next FRI layer given the values and domain of the current layer.
+    pub fn compute_next_fri_layer<F: FftField + PrimeField>(
+        domain: Radix2EvaluationDomain<F>,
+        input_layer: &[F],
+        eval_point: &F,
+    ) -> Result<Vec<F>, Error> {
+        assert_eq!(input_layer.len(), domain.size as usize);
+
+        let mut next_layer = Vec::with_capacity(input_layer.len() / 2);
+        for j in (0..input_layer.len()).step_by(2) {
+            let x_inv = get_field_element_at_index(&domain, j)
+                .inverse()
+                .ok_or_else(|| anyhow::anyhow!("Inverse not found"))?;
```

Comment:
Same comment about using batch inversion, or, just batch precompute all the "x_inv".
In the case of the domain, computing the inverse is pretty easy: the domain is a (coset of) a cyclic subgroup, i.e. so the inverse is just the element at index j + n/2.